include/h2o/stats.h

```diff
@@ -0,0 +1,105 @@
+#include <assert.h>
+#include <string.h>
+#include <stddef.h>
+#include <stdio.h>
+
+struct st_h2o_context_t;
+#define H2O_STATS_MAX_BUCKETS 100
+
+typedef struct st_h2o_stat_ops_t {
+    h2o_iovec_t name;
+    h2o_iovec_t desc;
+    size_t offset;
+    size_t nr_elements;
+    const char **labels;
+    void (*aggregate)(const struct st_h2o_stat_ops_t *self, struct st_h2o_context_t *agg, struct st_h2o_context_t *per_thread);
+    void (*stringify)(const struct st_h2o_stat_ops_t *self, struct st_h2o_context_t *agg, h2o_buffer_t **buf);
+    void (*prometheus)(const struct st_h2o_stat_ops_t *self, struct st_h2o_context_t *agg, h2o_buffer_t **buf);
+} h2o_stat_ops_t;
+
+typedef struct st_h2o_histogram_t {
+    struct {
+        uint64_t upper_bound;
+        uint64_t value;
+    } buckets[H2O_STATS_MAX_BUCKETS];
+    uint64_t inf;
+    uint64_t sum;
+    uint64_t count;
+} h2o_histogram_t;
+
+typedef struct st_h2o_metric_t {
+    uint64_t counter;
+    uint64_t gauge;
+} h2o_metric_t;
+
+static inline void h2o_histogram_init(h2o_histogram_t *hist, uint64_t start)
+{
+    size_t i;
+
+    memset(hist, 0, sizeof(*hist));
+
+    uint64_t interval = start;
+    for (i = 0; i < H2O_STATS_MAX_BUCKETS; i++) {
+        hist->buckets[i].upper_bound = interval;
+        if ((int)(interval * 0.1) <= 1)
+            interval++;
+        else
+            interval = interval + (int)(interval * 0.1);
+    }
+}
+
+static inline void h2o_histogram_add_one(h2o_histogram_t *hist, uint64_t value)
+{
+    size_t i;
+
+    hist->sum += value;
+    hist->count++;
+
+    for (i = 0; i < H2O_STATS_MAX_BUCKETS; i++) {
+        /* prometheus uses `le` for lower or equal, mimic this */
+        if (value <= hist->buckets[i].upper_bound) {
+            hist->buckets[i].value++;
+            return;
+        }
+    }
+    hist->inf++;
+}
+
+void hist_aggregate(const struct st_h2o_stat_ops_t *self, struct st_h2o_context_t *agg, struct st_h2o_context_t *per_thread);
+void hist_stringify(const struct st_h2o_stat_ops_t *self, struct st_h2o_context_t *agg, h2o_buffer_t **buf);
+void hist_prometheus(const struct st_h2o_stat_ops_t *self, struct st_h2o_context_t *agg, h2o_buffer_t **buf);
+void counter_stringify(const struct st_h2o_stat_ops_t *self, struct st_h2o_context_t *agg, h2o_buffer_t **buf);
+void counter_prometheus(const struct st_h2o_stat_ops_t *self, struct st_h2o_context_t *agg, h2o_buffer_t **buf);
+void gauge_stringify(const struct st_h2o_stat_ops_t *self, struct st_h2o_context_t *agg, h2o_buffer_t **buf);
+void gauge_prometheus(const struct st_h2o_stat_ops_t *self, struct st_h2o_context_t *agg, h2o_buffer_t **buf);
+void metric_aggregate(const struct st_h2o_stat_ops_t *self, struct st_h2o_context_t *agg, struct st_h2o_context_t *per_thread);
+void metric_stringify(const struct st_h2o_stat_ops_t *self, struct st_h2o_context_t *agg, h2o_buffer_t **buf);
+void metric_prometheus(const struct st_h2o_stat_ops_t *self, struct st_h2o_context_t *agg, h2o_buffer_t **buf);
```

Comment:
These functions should have `h2o_` prefix if we are to define them as extern.

The other way to proceed would be to define an aggregate of callbacks instead, e.g., `h2o_stat_op_type_t`, that contain number of function pointers. At the moment, in lib/handler/status/events.c, we define how a set of callbacks should be used together for each "type" (e.g., hist, metric, gauge, counter, counter_array). To paraphrase, the types are defined there while the member functions are defined here.